File: 0475. Heaters/main0475.py
from typing import List


class Solution0475:
    def findRadius(self, houses: List[int], heaters: List[int]) -> int:
        houses.sort()
        heaters.sort()
        
        m, n = len(houses), len(heaters)
        i, j = 0, 0
        heatRad = 0
        
        while i < m and j < n:
            curDist = abs(heaters[j] - houses[i])
            nextDist = float('inf')
            if j < n - 1:
                nextDist = abs(heaters[j+1] - houses[i])
            if curDist < nextDist:
                heatRad = max(heatRad, curDist)
                i += 1
            else:
                j += 1
        
        return heatRad


houses = [1,2,3]
heaters = [2]
sol = Solution0475()
rad = sol.findRadius(houses, heaters)
print(rad)

File: 0475. Heaters/main0475.java
public class main0475 {
    public static void main(String[] args) {
        Solution0475 sol = new Solution0475();
        int houses[], heaters[], res;

        houses = new int[]{1,2,3};
        heaters = new int[]{2};
        res = sol.findRadius(houses, heaters);
        System.out.println(res);

    }
}

class Solution0475 { // 30
    public int findRadius(int[] houses, int[] heaters) {
        Arrays.sort(houses);        
        Arrays.sort(heaters);
        int m = houses.length, n = heaters.length;
        int heatRad = 0;
        int i = 0, j = 0;
        while (i < m && j < n) {
            int curDist = Math.abs(houses[i] - heaters[j]);
            int nextDist = Integer.MAX_VALUE;
            if (j < n - 1) {
                nextDist = Math.abs(houses[i] - heaters[j+1]);
            }
            if (curDist < nextDist) {
                heatRad = Math.max(heatRad, curDist);
                i++;
            } else {
                j++;
            }
        }
        return heatRad;
    }
}